[CUSPARSE] Add sv! and sm! routines

amontoison opened the pull request:
`sv2` and `sm2` will be removed in the next major version of CUDA toolkit.

amontoison committed:
[CUSPARSE] Add sv! and sm! routines

amontoison committed:
Update libcusparse_common.jl

amontoison committed:
Update libcusparse.jl

amontoison committed:
Fix a typo in level3.jl

amontoison committed:
Add a missing end

amontoison commented:
`ERROR: LoadError: CUDA error: initialization error (code 3, ERROR_NOT_INITIALIZED)`
It's hard to know the origin of the problem :thinking: 

Solved: The `if CUSPARSE.version() < v"11.5.1"` must be inside functions.

amontoison committed:
Update generic.jl

amontoison committed:
Add descriptors in helpers.jl

amontoison committed:
Update generic.jl

amontoison committed:
Remove the variant of sv! and sm! for CuSparseMatrixCSC

amontoison committed:
Update generic.jl

amontoison committed:
Fix the initialization error

amontoison committed:
Use length instead of size for CuVector

amontoison commented:
@maleadt Do you idea what is the problem with these [lines](
https://github.com/JuliaGPU/CUDA.jl/pull/1593/files#diff-416606e9f5d8f57045482fc9fc855bccc8f31a27969e609160b17cc793501ae6R388-R389) ?

```julia
CUDA.CUSPARSE.cusparseSpMatGetAttribute(descA, CUSPARSE_SPMAT_FILL_MODE, 'L', Csize_t(4))

Internal error: encountered unexpected error during compilation of cusparseSpMatGetAttribute:
TypeError(func=:ccall, context="", expected=Symbol, got=CUDA.Deps.libcusparse)
...
ERROR: `ccall` requires the compiler
```

maleadt commented:
There an issue with the generated headers, you need to use `libcusparse()` in the `ccall`.

amontoison committed:
Update libcusparse.jl

amontoison committed:
Update generic.jl

amontoison committed:
Use CuPtr instead or Ptr for externalBuffer

amontoison committed:
Some some errors

amontoison committed:
Fix interfaces.jl

amontoison committed:
Add a missing Ref

amontoison committed:
Final fixes

amontoison commented:
Thanks @maleadt! The tests passed with the last version of CUDA toolkit.
Should we just test with the last version in `test/cusparse/generic.jl`?

I have the same issue with #1592.

maleadt commented:
What exactly is the issue? Can you find something relevant in the CUSPARSE changelogs? We can always make tests conditional, there's many like that already, but only if there's a good reason.

amontoison commented:
We have a lot of issues related to `cusparseSpSV`, `cusparseSpSM`, `cusparseSpMV` and `cusparseSpMM` in the CUSPARSE changelogs:
```
cuSPARSE: Release 11.7 Update 1]
    New Features
- Improved COO cusparseSpMM Alg2 with support for batched computation, custom row/col-major layout for B/C, and mixed-precision computation.
- Better performance for cusparseSpMM COO Alg3 and cusparseSpSM.
    Resolved Issues
- Batched cusparseSpMM produced wrong results when the number of columns of B/C is one.
    Known Issues
- cusparseSpSV, cusparseSpSM could produce wrong results if the output vector/matrix is not zero-initialized.
```
```
[cuSPARSE: Release 11.7]
    Resolved Issues
- Fixed a rare correctness bug of cusparseSpMM with CUSPARSE_SPMM_CSR_ALG1 when the number of rows in the sparse matrix is 2.
- cusparseSpSV, cusparseSpSM could produce wrong results if the output vector/matrix is not zero-initialized.

```
```
cuSPARSE: Release 11.6 Update 1

    New Features
- Improved CSR cusparseSpMM Alg1:
- Improved COO cusparseSpMM Alg3 with support for batched computation, custom row/col-major layout for B/C, and mixed-precision computation.
- Improved mixed-precision computation of CSR/COO cusparseSpMV.
- Added CSC format support for cusparseSpMV and cusparseSpMM.
    Known Issues
- cusparseSpSV, cusparseSpSM could produce wrong results if the output vector/matrix is not zero-initialized.
```
```
cuSPARSE: Release 11.6
    Resolved Issues
- Fixed incorrect results in COO SpMM Alg1 which occurred in some rare cases.
```
```
cuSPARSE: Release 11.5 Update 1
    Resolved Issues
- cusparseSpSM produced wrong results in some cases when the matB operation is CUSPARSE_ OPERATION_NON_TRANSPOSE or CUSPARSE_OPERATION_CONJUGATE_TRANSPOSE.
- cusparseSpSM produced wrong results in some cases when the matrix layout is row-major.
```
```
cuSPARSE: Release 11.4 Update 1
    Resolved Issues
- cusparseSpSV and cusparseSpSM could produce wrong results
- cusparseSpSV and cusparseSpSM did not work correctly when vecX == vecY or matB == matC.
```
```
cuSPARSE: Release 11.4
    Known Issues
- cusparseSpSV and cusparseSpSM could produce wrong results
- cusparseSpSV and cusparseSpSM do not work correctly when vecX == vecY or matB == matC.
```

I also submitted a bug to Nvidia yesterday:
- cusparseSpMV and cusparseSpMM: complex matrix-vector and matrix-matrix products with matrices stored CSC format and transa='C' return a wrong result (#1592) 
- cusparseSpSV and cusparseSpSM: forward and backward sweeps with complex triangular matrices, transa='C' and diag='N' return a wrong result (#1593)

amontoison committed:
Fix tests

amontoison committed:
Test sv and sm routines with CUSPARSE >= v11.7.0

amontoison commented:
@maleadt 
I have feedbacks from the CUSPARSE engineers:
- cusparseMV isn't supporting conj transpose for matrix A in CSC format at the moment. We'll add this feature in the next release.
- For cusparseSpSV and SpSM kernels: cusparse supports CUSPARSE_OPERATION_CONJUGATE_TRANSPOSE for CSR and COO. We found an issue that the diagonal of the sparse matrix A is not conjugated with CUSPARSE_OPERATION_CONJUGATE_TRANSPOSE. This is why the problem only shows for non-unit diagonals. This is a known issue and we will fix it in the next release.

I propose to wait the next release for #1592.
For this PR, I modified the tests to take into account the issue with SpSV and SpSM but we can also wait the next release.

maleadt commented on lib/cusparse/generic.jl:
```diff
@@ -57,11 +59,24 @@ Base.unsafe_convert(::Type{cusparseDnMatDescr_t}, desc::CuDenseMatrixDescriptor)
         cusparseCreateCsr(
             desc_ref,
-            size(A)..., length(nonzeros(A)),
+            size(A)..., A.nnz,
```
calling `nnz` would be even better.

maleadt commented:
Thanks for investigating!

>  For this PR, I modified the tests to take into account the issue with SpSV and SpSM but we can also wait the next release.

I'd be fine merging now, seeing how there's functionality/tests that are working already. Maybe add a note of open an issue to keep track of the functionality to enable for the next CUSPARSE release though?

amontoison committed:
Use nnz(A)

amontoison commented:
> Thanks for investigating!
> 
> > For this PR, I modified the tests to take into account the issue with SpSV and SpSM but we can also wait the next release.
> 
> I'd be fine merging now, seeing how there's functionality/tests that are working already. Maybe add a note of open an issue to keep track of the functionality to enable for the next CUSPARSE release though?

Yes, let's do that.
I opened the issue #1610.

amontoison commented on lib/cusparse/generic.jl:
```diff
@@ -345,3 +359,103 @@ function mm!(transa::SparseChar, transb::SparseChar, α::Number, A::CuSparseMatr
+    (mB != nC) && (transb != 'N') && throw(DimensionMismatch("B must have the same the number of rows that C has as columns, but B has $mB rows and C has $nC columns"))
+
+    if isa(A, CuSparseMatrixCSC)
+        # cusparseSpSM doesn't supports CSC format
```
```suggestion
        # cusparseSpSM doesn't support CSC format
```

amontoison commented on lib/cusparse/generic.jl:
```diff
@@ -345,3 +359,103 @@ function mm!(transa::SparseChar, transb::SparseChar, α::Number, A::CuSparseMatr
+    (mY != mA) && throw(DimensionMismatch("Y must have length $mA, but has length $mY"))
+
+    if isa(A, CuSparseMatrixCSC)
+        # cusparseSpSV doesn't supports CSC format
```
```suggestion
        # cusparseSpSV doesn't support CSC format
```

amontoison committed:
Update lib/cusparse/generic.jl

amontoison committed:
Update lib/cusparse/generic.jl

maleadt commented:
CI looks good, let's merge this. Thanks for the PR!

For future reference though, it may be better to keep the PR as 'draft' for as long as possible, to save on CI time when it is known not to be ready.
